python/ray/data/_internal/block_batching.py

```diff
@@ -119,74 +99,222 @@ def get_batches(block: Optional[ObjectRef[Block]] = None) -> Iterator[BatchType]
     else:
         prefetcher = WaitBlockPrefetcher()
 
-    # Batch blocks over the prefetch windows.
-    for block_window in _sliding_window(
-        blocks, prefetch_blocks + 1, clear_block_after_read
-    ):
-        block_window = list(block_window)
-        with stats.iter_wait_s.timer():
-            prefetcher.prefetch_blocks(block_window)
-        yield from get_batches(block_window[0])
-
-    # Consume remainder of final block window.
-    for block in block_window[1:]:
-        yield from get_batches(block)
-
-    # Consume any remaining batches, now that we're done adding blocks to the batcher.
-    yield from get_batches()
-
-
-def _format_batch(batch: Block, batch_format: str) -> BatchType:
-    if batch_format == "default" or batch_format == "native":
-        batch = BlockAccessor.for_block(batch).to_default()
-    elif batch_format == "pandas":
-        batch = BlockAccessor.for_block(batch).to_pandas()
-    elif batch_format == "pyarrow":
-        batch = BlockAccessor.for_block(batch).to_arrow()
-    elif batch_format == "numpy":
-        batch = BlockAccessor.for_block(batch).to_numpy()
-    else:
-        raise ValueError(
-            f"The given batch format '{batch_format}' is invalid. Supported "
-            f"`batch_format` values: {'default', 'pandas', 'pyarrow', 'numpy'}."
-        )
-    return batch
+    block_iter = _resolve_blocks(
+        _prefetch_blocks(
+            block_ref_iter=block_refs,
+            prefetcher=prefetcher,
+            stats=stats,
+            num_blocks_to_prefetch=prefetch_blocks,
+            clear_block_after_read=clear_block_after_read,
+        ),
+        stats=stats,
+    )
+
+    yield from batch_blocks(
+        block_iter,
+        stats=stats,
+        batch_size=batch_size,
+        batch_format=batch_format,
+        drop_last=drop_last,
+        shuffle_buffer_min_size=shuffle_buffer_min_size,
+        shuffle_seed=shuffle_seed,
+        ensure_copy=ensure_copy,
+    )
+
+
+def batch_blocks(
+    blocks: Iterator[Block],
+    stats: Optional[Union[DatasetStats, DatasetPipelineStats]] = None,
```

Comment:
as this parameter becomes optional, should it be put under `*`?

Reply:
updated!